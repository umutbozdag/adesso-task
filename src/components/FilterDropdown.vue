<template>
  <div class="z-10 inline-block" v-on-click-outside="closeDropdown">
    <button
      @click="isOpen = !isOpen"
      class="justify-center w-full px-4 py-2 text-sm font-medium text-white bg-gray-900 border border-gray-600 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-blue-500"
    >
      {{ selectedOptionLabel }}: {{ selectedOption }}
    </button>
    <div
      v-if="isOpen"
      class="origin-top-right absolute mt-2 w-48 rounded-md shadow-lg z-10 bg-gray-900 ring-1 ring-black ring-opacity-5"
    >
      <div class="py-2 p-2" role="menu">
        <a
          v-for="option in options"
          :key="option.name"
          class="flex rounded-md px-4 py-2 text-sm text-white hover:bg-gray-800 cursor-pointer"
          role="menuitem"
          @click="handleOptionClick(option)"
        >
          {{ option.label }}
        </a>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { UnitAgeEnum } from '@/types/Unit';
import { FilterDropdownOption, FilterDropdownEmitEnum } from '@/types/FilterDropdown';
import { ref } from 'vue';
import { vOnClickOutside } from '@vueuse/components';

interface IFilterDropdownProps {
  options: FilterDropdownOption[];
  selectedOption: UnitAgeEnum;
  selectedOptionLabel: string;
}

defineProps<IFilterDropdownProps>();

const emit = defineEmits<{
  (event: 'option-selected', option: FilterDropdownOption): void;
}>();

const isOpen = ref(false);

const closeDropdown = () => {
  isOpen.value = false;
};

const handleOptionClick = (option: FilterDropdownOption) => {
  emit(FilterDropdownEmitEnum.OPTION_SELECTED, option);
  closeDropdown();
};
</script>
@/types/FilterDropdown
