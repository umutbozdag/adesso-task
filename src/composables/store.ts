import RootState from '@/types/store/State';
import { InjectionKey } from 'vue';
import { Store, useStore as baseUseStore } from 'vuex';

export const key: InjectionKey<Store<RootState>> = Symbol();

export function useStore(): Store<RootState> {
  return baseUseStore(key);
}
