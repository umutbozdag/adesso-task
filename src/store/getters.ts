import { UnitAgeEnum } from '@/types/Unit';
import { IUnitDTO } from '@/types/dto/UnitDTO';
import Getters, { GetterTypes } from '@/types/store/Getters';
import RootState from '@/types/store/State';
import { GetterTree } from 'vuex';

const getters: GetterTree<RootState, RootState> & Getters = {
  [GetterTypes.UNITS]: (state: RootState) => state.units,
  [GetterTypes.SELECTED_OPTION]: (state) => state.selectedOption,
  [GetterTypes.CURRENT_UNITS_DATA]: (state) =>
    (state.units as Record<UnitAgeEnum, IUnitDTO[]>)[state.selectedOption]
};

export default getters;
