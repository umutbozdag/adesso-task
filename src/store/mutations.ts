import RootState from '@/types/store/State';
import { MutationTree } from 'vuex';
import Mutations, { MutationTypes } from '@/types/store/Mutations';
import { CostFilterEnum } from '@/types/UnitTableFilter';

const mutations: MutationTree<RootState> & Mutations = {
  [MutationTypes.SET_SELECTED_OPTION](state, option) {
    state.selectedOption = option;
  },
  [MutationTypes.SET_UNITS](state, { units, key }) {
    state.units[key] = units;
  },
  [MutationTypes.SET_FILTERED_UNITS](state, units) {
    state.filteredUnits = units;
  },
  [MutationTypes.SET_UNIT_DETAIL](state, unit) {
    state.unitDetail = unit;
  },
  [MutationTypes.RESET_COST_FILTERS](state) {
    Object.keys(state.costsFilters).forEach((key) => {
      state.costsFilters[key as CostFilterEnum].selected = false;
      state.costsFilters[key as CostFilterEnum].sliderValue = '0';
    });
  }
};
export default mutations;
