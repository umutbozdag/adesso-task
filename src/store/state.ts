import { UnitAgeEnum } from '@/types/Unit';
import { CostFilterEnum } from '@/types/UnitTableFilter';
import RootState from '@/types/store/State';

const state: RootState = {
  selectedOption: UnitAgeEnum.All,
  units: {
    [UnitAgeEnum.All]: null,
    [UnitAgeEnum.Castle]: null,
    [UnitAgeEnum.Dark]: null,
    [UnitAgeEnum.Feudal]: null,
    [UnitAgeEnum.Imperial]: null
  },
  filteredUnits: [],
  unitDetail: null,
  costsFilters: {
    [CostFilterEnum.WOOD]: {
      label: 'Wood',
      sliderValue: '0',
      selected: false
    },
    [CostFilterEnum.FOOD]: {
      label: 'Food',
      sliderValue: '0',
      selected: false
    },
    [CostFilterEnum.GOLD]: {
      label: 'Gold',
      sliderValue: '0',
      selected: false
    }
  }
};

export default state;
