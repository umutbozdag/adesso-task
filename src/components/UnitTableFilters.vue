<template>
  <h3 class="text-white text-lg font-bold mb-8">Filters</h3>
  <div class="flex items-center">
    <div class="text-white text-md font-semibold mb-4 w-1/6">Ages</div>
    <div class="text-white text-md font-semibold mb-4">Costs</div>
  </div>
  <div class="flex items-baseline">
    <div class="w-1/6">
      <filter-dropdown
        :options="filterOptionsConfig"
        @option-selected="handleOptionSelected"
        :selected-option="selectedOption"
        :selected-option-label="dropdownSelectedOptionLabel"
      />
    </div>

    <span
      class="flex items-center"
      v-for="(filter, filterIndex) in costsFilters"
      :key="filterIndex"
    >
      <range-slider v-model="filter.sliderValue" :disabled="!filter.selected">
        <filter-checkbox v-model="filter.selected" :key="filterIndex" />
        <template #slider-label> {{ filter.label }} </template>
      </range-slider>
    </span>
  </div>
</template>

<script setup lang="ts">
import { useStore } from '@/composables/store';
import FilterDropdown from '@/components/FilterDropdown.vue';
import { filterOptionsConfig } from '@/config/filter';
import { FilterDropdownOption } from '@/types/FilterDropdown';
import { computed, watch, nextTick } from 'vue';
import RangeSlider from '@/components/RangeSlider.vue';
import { ActionTypes } from '@/types/store/Actions';
import FilterCheckbox from './FilterCheckbox.vue';
import { CostFilterConfigType } from '@/types/UnitTableFilter';

const store = useStore();

const emit = defineEmits<{
  (e: 'filters-changed', filters: CostFilterConfigType): void;
}>();

const costsFilters = computed(() => store.state.costsFilters);
const dropdownSelectedOptionLabel = 'Ages';

watch(
  costsFilters,
  (newCostsFilter) => {
    emit('filters-changed', newCostsFilter);
  },
  {
    deep: true
  }
);

const selectedOption = computed(() => store.getters.selectedOption);

const handleOptionSelected = async (option: FilterDropdownOption) => {
  store.dispatch(ActionTypes.RESET_COST_FILTERS);
  await nextTick();
  store.dispatch(ActionTypes.SET_FILTERED_UNITS, []);
  store.dispatch(ActionTypes.SET_SELECTED_OPTION, option.name);
};
</script>
