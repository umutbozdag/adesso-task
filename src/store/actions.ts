import { ActionTree } from 'vuex';
import Actions, { ActionTypes } from '@/types/store/Actions';
import RootState from '@/types/store/State';
import { MutationTypes } from '@/types/store/Mutations';
import { IUnits, UnitAgeEnum } from '@/types/Unit';
import UnitDto from '@/dto/UnitDTO';
import { IUnitDTO } from '@/types/dto/UnitDTO';
import { CostFilterEnum } from '@/types/UnitTableFilter';
import filterByCost from '@/utils/filterByCost';

const actions: ActionTree<RootState, RootState> & Actions = {
  async [ActionTypes.FETCH_UNITS]({ commit }) {
    try {
      const response: Response = await fetch(
        `${import.meta.env.VITE_LOCALHOST_URL}/age-of-empires-units.json`
      );
      const data: IUnits = await response.json();

      const dto = new UnitDto(data);
      commit(MutationTypes.SET_UNITS, {
        units: dto.units,
        key: UnitAgeEnum.All
      });
    } catch (e) {
      console.error(e);
    }
  },
  [ActionTypes.SET_SELECTED_OPTION]({ commit, dispatch }, option) {
    commit(MutationTypes.SET_SELECTED_OPTION, option);
    dispatch(ActionTypes.FILTER_BY_AGE, option);
  },
  [ActionTypes.FILTER_BY_AGE]({ commit, state }, age) {
    if (age === UnitAgeEnum.All) {
      return;
    }

    // TODO: Add type for Record<UnitAgeEnum, IUnitDTO[]>
    // If it's not filtered before
    if (!state.units[age]) {
      const filteredUnitsByAge = (state.units as Record<UnitAgeEnum, IUnitDTO[]>)[
        UnitAgeEnum.All
      ].filter((unit) => unit.age === age);
      commit(MutationTypes.SET_UNITS, {
        units: filteredUnitsByAge,
        key: age
      });
    }
  },
  [ActionTypes.FILTER_BY_COST]({ state, dispatch }, costsFilters) {
    const currentUnitsData = state.units[state.selectedOption] as IUnitDTO[];

    const costsUnitsData = currentUnitsData.filter(
      (unit: IUnitDTO) => unit.costs && unit.costs.length
    );

    const costFilters = [
      {
        name: CostFilterEnum.FOOD,
        prop: 'Food'
      },
      {
        name: CostFilterEnum.WOOD,
        prop: 'Wood'
      },
      {
        name: CostFilterEnum.GOLD,
        prop: 'Gold'
      }
    ];

    const currentFilteredData = costsUnitsData.filter((unit) => {
      const filterConditions: boolean[] = [];

      costFilters.forEach((filter) => {
        filterByCost(filter.prop, costsFilters[filter.name], unit, filterConditions);
      });

      return filterConditions.every((condition) => condition);
    });
    dispatch(ActionTypes.SET_FILTERED_UNITS, currentFilteredData);
  },
  [ActionTypes.SET_FILTERED_UNITS]({ commit }, units) {
    commit(MutationTypes.SET_FILTERED_UNITS, units);
  },
  async [ActionTypes.FETCH_UNIT_BY_ID]({ commit, state }, unitId) {
    const units = state.units[state.selectedOption];

    const unitDetail = units?.find((unit) => unit.id === unitId);
    commit(MutationTypes.SET_UNIT_DETAIL, unitDetail);
  },
  [ActionTypes.RESET_COST_FILTERS]({ commit }) {
    commit(MutationTypes.RESET_COST_FILTERS);
  }
};

export default actions;
