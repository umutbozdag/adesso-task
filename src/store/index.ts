import { Store, createStore } from 'vuex';
import actions from './actions';
import RootState from '@/types/store/State';
import state from './state';
import mutations from './mutations';
import getters from './getters';

const store = createStore<RootState>({
  state,
  mutations,
  actions,
  getters
});

export default store as Store<RootState>;
