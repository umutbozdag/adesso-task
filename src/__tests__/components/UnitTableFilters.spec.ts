import UnitTableFilters from '@/components/UnitTableFilters.vue';
import { VueWrapper, mount } from '@vue/test-utils';
import { expect, it, describe, beforeAll, vi, SpyInstance } from 'vitest';
import { ComponentPublicInstance } from 'vue';
import { createStore } from 'vuex';
import state from '@/store/state';
import actions from '@/store/actions';
import mutations from '@/store/mutations';
import getters from '@/store/getters';
import { FilterDropdownOption } from '@/types/FilterDropdown';
import { key } from '@/composables/store';
import { generateArrayOfUnits } from '../mocks/unit.mocks';

describe('UnitTableFilters', () => {
  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>;
  const store = createStore({
    state,
    actions,
    mutations,
    getters
  });
  let dispatchSpy: SpyInstance;
  beforeAll(() => {
    dispatchSpy = vi.spyOn(store, 'dispatch');
    wrapper = mount(UnitTableFilters, {
      global: {
        plugins: [[store, key]]
      }
    });
  });

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('handleOptionSelected dispatching SET_SELECTED_OPTION', async () => {
    const unitsMock = generateArrayOfUnits(1);
    store.state.units.All = unitsMock;
    const option: FilterDropdownOption = {
      label: 'Option',
      name: 'name'
    };
    wrapper.vm.handleOptionSelected(option);
    expect(dispatchSpy).toHaveBeenCalled();
  });
});
