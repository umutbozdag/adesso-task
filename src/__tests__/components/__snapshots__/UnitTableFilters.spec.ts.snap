// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnitTableFilters > renders correctly 1`] = `
"<h3 class=\\"text-white text-lg font-bold mb-8\\">Filters</h3>
<div class=\\"flex items-center\\">
  <div class=\\"text-white text-md font-semibold mb-4 w-1/6\\">Ages</div>
  <div class=\\"text-white text-md font-semibold mb-4\\">Costs</div>
</div>
<div class=\\"flex items-baseline\\">
  <div class=\\"w-1/6\\">
    <div class=\\"z-10 inline-block\\"><button class=\\"justify-center w-full px-4 py-2 text-sm font-medium text-white bg-gray-900 border border-gray-600 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-blue-500\\">Ages: All</button>
      <!--v-if-->
    </div>
  </div><span class=\\"flex items-center\\"><div class=\\"flex items-center\\"><div class=\\"flex items-center mr-1\\"><input type=\\"checkbox\\" class=\\"w-4 h-4 rounded focus:ring-blue-600 ring-offset-gray-800 focus:ring-2 bg-gray-700 border-gray-600\\" value=\\"false\\"><!--v-if--></div><label class=\\"text-sm font-medium text-white mr-2\\">Wood</label><input type=\\"range\\" min=\\"0\\" max=\\"200\\" disabled=\\"\\" class=\\"w-full h-2 rounded-lg appearance-none cursor-pointer bg-gray-700\\"></div><!--v-if--></span><span class=\\"flex items-center\\"><div class=\\"flex items-center\\"><div class=\\"flex items-center mr-1\\"><input type=\\"checkbox\\" class=\\"w-4 h-4 rounded focus:ring-blue-600 ring-offset-gray-800 focus:ring-2 bg-gray-700 border-gray-600\\" value=\\"false\\"><!--v-if--></div><label class=\\"text-sm font-medium text-white mr-2\\">Food</label><input type=\\"range\\" min=\\"0\\" max=\\"200\\" disabled=\\"\\" class=\\"w-full h-2 rounded-lg appearance-none cursor-pointer bg-gray-700\\"></div><!--v-if--></span><span class=\\"flex items-center\\"><div class=\\"flex items-center\\"><div class=\\"flex items-center mr-1\\"><input type=\\"checkbox\\" class=\\"w-4 h-4 rounded focus:ring-blue-600 ring-offset-gray-800 focus:ring-2 bg-gray-700 border-gray-600\\" value=\\"false\\"><!--v-if--></div><label class=\\"text-sm font-medium text-white mr-2\\">Gold</label><input type=\\"range\\" min=\\"0\\" max=\\"200\\" disabled=\\"\\" class=\\"w-full h-2 rounded-lg appearance-none cursor-pointer bg-gray-700\\"></div><!--v-if--></span>
</div>"
`;
